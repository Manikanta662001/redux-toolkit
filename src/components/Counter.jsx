import React from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  increment,
  decrement,
  reset,
  incrementAsync,
} from "../store/slices/counterSlice";
import { persistor } from "../store/store";
function Counter() {
  const dispatch = useDispatch();
  const count = useSelector((state) => state.counterReducer?.count);
  
  const handleReset = ()=>{
    dispatch({type:"RESET"});
    persistor.purge();
  }
  return (
    <div>
      <h5>{count}</h5>
      <button onClick={() => dispatch(increment())}>+</button>
      <button onClick={() => dispatch(decrement())}>-</button>
      <button onClick={() => dispatch(reset())}>reset</button>
      <button onClick={() => dispatch(incrementAsync())}>+ async</button>
      <button onClick={() => window.location.reload()}>Reload</button>
      <button onClick={() => handleReset()}>Empty Store</button>
    </div>
  );
}

export default Counter;
