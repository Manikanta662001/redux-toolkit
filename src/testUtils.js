import { combineReducers, configureStore } from "@reduxjs/toolkit";
import { fetchUsers } from "./store/thunks/fetchUsersThunk";
import counterReducer from "./store/slices/counterSlice";
import apiReducer from "./store/slices/apiSlice";
import { render } from "@testing-library/react";
import storage from "redux-persist/lib/storage";
import { persistReducer } from "redux-persist";
import { Provider } from "react-redux";

export const renderWithProvider = (
  ui,
  preloadedState = {},
  customReducers = {}
) => {
  const rootReducer = combineReducers({
    apiReducer,
    counterReducer,
    ...customReducers,
  });
  const rootResetReducer = (state, action) => {
    if (action.type === "RESET") {
      return {
        ...state,
        counterReducer: undefined,
        apiReducer: undefined, // Reset this part of the state
      };
    }
    return rootReducer(state, action);
  };
  const persistConfig = {
    key: "root",
    storage,
    whitelist: ["counterReducer"],
  };

  const persistedReducer = persistReducer(persistConfig, rootResetReducer);

  const store = configureStore({
    reducer: persistedReducer,
    extraReducers: (builder) => {
      builder.addCase(fetchUsers.pending, (state) => {
        return { ...state, isLoading: true };
      });
      builder.addCase(fetchUsers.fulfilled, (state, action) => {
        return { ...state, isLoading: false, users: action.payload };
      });
      builder.addCase(fetchUsers.rejected, (state, action) => {
        return { ...state, isLoading: false, error: action.payload };
      });
    },
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        serializableCheck: {
          ignoredActions: [
            "persist/PERSIST",
            "persist/REHYDRATE",
            "persist/PAUSE",
            "persist/PURGE",
            "persist/FLUSH",
            "persist/REGISTER",
          ],
        },
      }),
    preloadedState,
    devTools: true,
  });
  if (ui === undefined) return store;
  return { ...render(<Provider store={store}>{ui}</Provider>), store };
};
