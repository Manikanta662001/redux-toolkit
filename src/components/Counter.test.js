import { screen, cleanup, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithProvider } from "../testUtils";
import Counter from "./Counter";

describe("Counter Component", () => {
  const initialState = {
    counterReducer: { count: 0 },
    apiReducer: {
      users: [],
      isLoading: false,
      error: "",
    },
  };

  test("renders correctly and Match the Snapshot", () => {
    const { asFragment } = renderWithProvider(<Counter />, initialState);
    expect(asFragment(<Counter />)).toMatchSnapshot();
    expect(screen.getByRole("heading", { level: 5 })).toHaveTextContent("0");
  });
  afterEach(cleanup);

  test("dispatches increment action when we click + button", async () => {
    const { store } = renderWithProvider(<Counter />, initialState);
    const incButton = screen.getByRole("button", { name: "+" });
    await userEvent.click(incButton);
    expect(screen.getByRole("heading", { level: 5 })).toHaveTextContent("1");
    //to test the store value
    console.log("STATE::::", store.getState());
    expect(store.getState().counterReducer.count).toBe(1);
  });
  test("dispatches decrement action when we click - button", async () => {
    const { store } = renderWithProvider(<Counter />, initialState);
    const decButton = screen.getByRole("button", { name: "-" });
    await userEvent.click(decButton);
    expect(screen.getByRole("heading", { level: 5 })).toHaveTextContent("-1");
    //to test the store value
    expect(store.getState().counterReducer.count).toBe(-1);
  });
  test("dispatches reset action when we click RESET button", async () => {
    const { store } = renderWithProvider(<Counter />, initialState);
    const resetButton = screen.getByRole("button", { name: "reset" });
    await userEvent.click(resetButton);
    expect(screen.getByRole("heading", { level: 5 })).toHaveTextContent("0");
    //to test the store value
    expect(store.getState().counterReducer.count).toBe(0);
  });
  test("increments by some delay", async () => {
    const { store } = renderWithProvider(<Counter />, initialState);
    const delayButton = screen.getByRole("button", { name: "+ async" });
    await userEvent.click(delayButton);
    await waitFor(
      () => {
        expect(store.getState().counterReducer.count).toBe(1);
        expect(screen.getByRole("heading", { level: 5 })).toHaveTextContent(
          "1"
        );
      },
      { timeout: 2000 }
    );
  });

  test("If we reload also count should not reset", () => {
    const { store } = renderWithProvider(<Counter />, {
      counterReducer: { count: 10 },
    });
    expect(store.getState().counterReducer.count).toBe(10);
    const reloadBtn = screen.getByRole("button", { name: "Reload" });
    userEvent.click(reloadBtn);
    expect(store.getState().counterReducer.count).toBe(10);
  });
  test("RESET action resets the store state", () => {
    const { store } = renderWithProvider(<Counter />, initialState);
    expect(store.getState().counterReducer.count).toBe(0);
    expect(store.getState().apiReducer.users).toStrictEqual([]);
    const emptyBtn = screen.getByRole("button", { name: "Empty Store" });
    userEvent.click(emptyBtn);
    expect(store.getState().counterReducer).toBeUndefined();
    expect(store.getState().apiReducer).toBeUndefined();
  });
});
