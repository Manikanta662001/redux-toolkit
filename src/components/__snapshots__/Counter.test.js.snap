// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component renders correctly and Match the Snapshot 1`] = `
<DocumentFragment>
  <div>
    <h5>
      0
    </h5>
    <button>
      +
    </button>
    <button>
      -
    </button>
    <button>
      reset
    </button>
    <button>
      + async
    </button>
    <button>
      Reload
    </button>
    <button>
      Empty Store
    </button>
  </div>
</DocumentFragment>
`;
